File: src/tutorial/ts/namespace/baseTypes-external.ts
// shows how to use contract external base types

/// <reference path="../../../main/ts/namespace/yass"/>

interface Integer {
    get(): number;
}

class IntegerHandler implements yass.TypeHandler<Integer> {
    read(reader: yass.Reader): Integer {
        return new IntegerImpl(reader.readZigZagInt());
    }
    write(value: Integer, writer: yass.Writer): void {
        writer.writeZigZagInt(value.get());
    }
    static TYPE_DESC = new yass.TypeDesc(yass.FIRST_ID, new IntegerHandler);
}

class IntegerImpl implements Integer {
    constructor(private value: number) {
        // note: check if value is really a Java Integer should be implemented here
    }
    get(): number {
        return this.value;
    }
    toString(): string {
        return "IntegerImpl(" + this.get() + ")";
    }
}

(<any>IntegerImpl).TYPE_DESC = IntegerHandler.TYPE_DESC;

File: py3/tutorial/base_types_external.py
# shows how to use contract external base types

import yass


class IntegerHandler(yass.BaseTypeHandler):
    def readBase(self, reader: yass.Reader) -> 'Integer':
        return Integer(reader.readZigZagInt())

    def writeBase(self, value: 'Integer', writer: yass.Writer) -> None:
        writer.writeZigZagInt(value.value)


class Integer:
    TYPE_DESC = yass.TypeDesc(yass.FIRST_DESC_ID, IntegerHandler())

    def __init__(self, value: int) -> None:
        # note: check if value is really a Java Integer should be implemented here
        self.value = value

    def __str__(self) -> str:
        return str(self.value)
